#include "MainWindow.h"
#include "./ui_mainwindow.h"
#include <QMessageBox>
#include <QPushButton>
#include <QLabel>
#include <QScrollArea>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , recordingActive(false)
    , currentTargetString("")
    , currentTargetFrequency(0.0f)
    , manualStringSelection(false)
{
    ui->setupUi(this);

    audioRecorder = new QtAudioRecorder(this);

    // –ü–æ–¥–∫–ª—é—á–∞–µ–º —Å–∏–≥–Ω–∞–ª –æ –æ–±–Ω–∞—Ä—É–∂–µ–Ω–Ω–æ–π –≤—ã—Å–æ—Ç–µ —Ç–æ–Ω–∞
    connect(audioRecorder, &QtAudioRecorder::pitchDetected,
            this, &MainWindow::updateTunerDisplay, Qt::QueuedConnection);

    connect(audioRecorder, &QtAudioRecorder::errorOccurred,
            this, &MainWindow::handleAudioError);

    // –ü–æ–¥–∫–ª—é—á–∞–µ–º –∫–Ω–æ–ø–∫–∏ —Å—Ç—Ä—É–Ω
    connect(ui->e2Button, &QPushButton::clicked, this, &MainWindow::onE2ButtonClicked);
    connect(ui->aButton, &QPushButton::clicked, this, &MainWindow::onAButtonClicked);
    connect(ui->dButton, &QPushButton::clicked, this, &MainWindow::onDButtonClicked);
    connect(ui->gButton, &QPushButton::clicked, this, &MainWindow::onGButtonClicked);
    connect(ui->bButton, &QPushButton::clicked, this, &MainWindow::onBButtonClicked);
    connect(ui->e4Button, &QPushButton::clicked, this, &MainWindow::onE4ButtonClicked);

    connect(ui->helpButton, &QPushButton::clicked, this, &MainWindow::showHelpDialog);

    connect(ui->autoButton, &QPushButton::clicked, this, [this]() {
        manualStringSelection = false;
        currentTargetString = "";
        currentTargetFrequency = 0.0f;
        updateTargetIndicator();
        resetStringHighlights();
        ui->statusbar->showMessage("Auto detection enabled", 2000);
    });

    // –ù–∞—á–∞–ª—å–Ω—ã–µ –Ω–∞—Å—Ç—Ä–æ–π–∫–∏ UI
    ui->startStopButton->setText("üé§ –°—Ç–∞—Ä—Ç");
    resetDisplay();

    // –î–æ–±–∞–≤–ª—è–µ–º –∏–Ω–¥–∏–∫–∞—Ç–æ—Ä —Ü–µ–ª–µ–≤–æ–π –Ω–æ—Ç—ã –≤ —Å—Ç–∞—Ç—É—Å–±–∞—Ä
    QLabel *targetIndicator = new QLabel("–†–µ–∂–∏–º: –ê–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏–π", this);
    targetIndicator->setObjectName("targetIndicator");
    targetIndicator->setStyleSheet("color: #4facfe; font-weight: bold;");
    ui->statusbar->addPermanentWidget(targetIndicator);
}

MainWindow::~MainWindow()
{
    if (recordingActive) {
        audioRecorder->stopRecording();
    }
    delete audioRecorder;
    delete ui;
}


void MainWindow::setTargetString(const QString& stringName, float frequency)
{
    currentTargetString = stringName;
    currentTargetFrequency = frequency;
    manualStringSelection = true;

    // –û–±–Ω–æ–≤–ª—è–µ–º UI
    updateTargetIndicator();

    // –°–±—Ä–∞—Å—ã–≤–∞–µ–º –≤—ã–¥–µ–ª–µ–Ω–∏–µ –≤—Å–µ—Ö –∫–Ω–æ–ø–æ–∫
    resetStringHighlights();

    // –í—ã–¥–µ–ª—è–µ–º –≤—ã–±—Ä–∞–Ω–Ω—É—é —Å—Ç—Ä—É–Ω—É
    if (stringName == "E2") {
        ui->e2Button->setChecked(true);
        ui->e2Button->setStyleSheet(ui->e2Button->styleSheet() +
                                    "QPushButton:checked { background-color: rgba(255, 107, 107, 0.8); }");
    } else if (stringName == "A") {
        ui->aButton->setChecked(true);
        ui->aButton->setStyleSheet(ui->aButton->styleSheet() +
                                   "QPushButton:checked { background-color: rgba(77, 171, 247, 0.8); }");
    } else if (stringName == "D") {
        ui->dButton->setChecked(true);
        ui->dButton->setStyleSheet(ui->dButton->styleSheet() +
                                   "QPushButton:checked { background-color: rgba(106, 255, 182, 0.8); }");
    } else if (stringName == "G") {
        ui->gButton->setChecked(true);
        ui->gButton->setStyleSheet(ui->gButton->styleSheet() +
                                   "QPushButton:checked { background-color: rgba(255, 218, 121, 0.8); }");
    } else if (stringName == "B") {
        ui->bButton->setChecked(true);
        ui->bButton->setStyleSheet(ui->bButton->styleSheet() +
                                   "QPushButton:checked { background-color: rgba(200, 121, 255, 0.8); }");
    } else if (stringName == "E4") {
        ui->e4Button->setChecked(true);
        ui->e4Button->setStyleSheet(ui->e4Button->styleSheet() +
                                    "QPushButton:checked { background-color: rgba(255, 121, 177, 0.8); }");
    }

    // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º —Å–æ–æ–±—â–µ–Ω–∏–µ
    ui->statusbar->showMessage(QString("–†–µ–∂–∏–º: %1 (%2 –ì—Ü)").arg(stringName).arg(frequency), 3000);
}

void MainWindow::updateTargetIndicator()
{
    QLabel *indicator = ui->statusbar->findChild<QLabel*>("targetIndicator");
    if (indicator) {
        if (manualStringSelection && !currentTargetString.isEmpty()) {
            indicator->setText(QString("–†–µ–∂–∏–º: %1").arg(currentTargetString));
            indicator->setStyleSheet("color: #ff9a9e; font-weight: bold;");
        } else {
            indicator->setText("–†–µ–∂–∏–º: –ê–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏–π");
            indicator->setStyleSheet("color: #4facfe; font-weight: bold;");
        }
    }
}

void MainWindow::highlightCorrectString(float pitchHz)
{
    // –ï—Å–ª–∏ —Ä—É—á–Ω–æ–π –≤—ã–±–æ—Ä —Å—Ç—Ä—É–Ω—ã –≤–∫–ª—é—á–µ–Ω, –Ω–µ –º–µ–Ω—è–µ–º –∞–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏
    if (manualStringSelection) {
        // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º –æ—Ç–∫–ª–æ–Ω–µ–Ω–∏–µ –æ—Ç –≤—ã–±—Ä–∞–Ω–Ω–æ–π —Å—Ç—Ä—É–Ω—ã
        if (currentTargetFrequency > 0) {
            float cents = 1200.0f * std::log2(pitchHz / currentTargetFrequency);

            // –û–±–Ω–æ–≤–ª—è–µ–º –ø—Ä–æ–≥—Ä–µ—Å—Å-–±–∞—Ä –æ—Ç–Ω–æ—Å–∏—Ç–µ–ª—å–Ω–æ —Ü–µ–ª–µ–≤–æ–π —á–∞—Å—Ç–æ—Ç—ã
            int barValue = qBound(-50, qRound(cents), 50);
            ui->tuningBar->setValue(barValue);

            // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º —Ü–µ–ª–µ–≤—É—é —á–∞—Å—Ç–æ—Ç—É –≤ –∏–Ω—Ç–µ—Ä—Ñ–µ–π—Å–µ
            ui->noteLabel->setText(QString("%1 ‚ûî %2 –ì—Ü").arg(currentTargetString).arg(currentTargetFrequency, 0, 'f', 1));
        }
        return;
    }

    // –ê–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏–π –ø–æ–∏—Å–∫ –±–ª–∏–∂–∞–π—à–µ–π —Å—Ç—Ä—É–Ω—ã
    QString closestString;
    float minDiff = 9999;

    for (auto it = stringFrequencies.begin(); it != stringFrequencies.end(); ++it) {
        float diff = qAbs(pitchHz - it.value());
        if (diff < minDiff) {
            minDiff = diff;
            closestString = it.key();
        }
    }

    // –°–±—Ä–æ—Å –≤—Å–µ—Ö —Å—Ç—Ä—É–Ω
    resetStringHighlights();

    // –ü–æ–¥—Å–≤–µ—Ç–∫–∞ –ø—Ä–∞–≤–∏–ª—å–Ω–æ–π —Å—Ç—Ä—É–Ω—ã, –µ—Å–ª–∏ –¥–æ—Å—Ç–∞—Ç–æ—á–Ω–æ –±–ª–∏–∑–∫–æ
    if (minDiff < 10) {
        currentTargetString = closestString;
        currentTargetFrequency = stringFrequencies[closestString];

        if (closestString == "E2") {
            ui->e2Button->setChecked(true);
        } else if (closestString == "A") {
            ui->aButton->setChecked(true);
        } else if (closestString == "D") {
            ui->dButton->setChecked(true);
        } else if (closestString == "G") {
            ui->gButton->setChecked(true);
        } else if (closestString == "B") {
            ui->bButton->setChecked(true);
        } else if (closestString == "E4") {
            ui->e4Button->setChecked(true);
        }

        updateTargetIndicator();
    }
}

void MainWindow::resetStringHighlights()
{
    ui->e2Button->setChecked(false);
    ui->aButton->setChecked(false);
    ui->dButton->setChecked(false);
    ui->gButton->setChecked(false);
    ui->bButton->setChecked(false);
    ui->e4Button->setChecked(false);
}

void MainWindow::resetDisplay()
{
    ui->frequencyLabel->setText("--- –ì—Ü");
    ui->frequencyLabel->setStyleSheet("color: #00dbde; background: transparent;");
    ui->noteLabel->setText("---");
    ui->noteLabel->setStyleSheet("color: #ff9a9e; background: transparent;");
    ui->centsLabel->setText("–¶–µ–Ω—Ç—ã: ---");
    ui->centsLabel->setStyleSheet("color: #a0aec0; background: transparent;");
    ui->tuningBar->setValue(0);

    // –°–±—Ä–æ—Å —Ü–µ–ª–µ–≤–æ–π —Å—Ç—Ä—É–Ω—ã
    currentTargetString = "";
    currentTargetFrequency = 0.0f;
    manualStringSelection = false;
    updateTargetIndicator();
    resetStringHighlights();
}


void MainWindow::onE2ButtonClicked()
{
    setTargetString("E2", 82.41f);
}

void MainWindow::onAButtonClicked()
{
    setTargetString("A", 110.00f);
}

void MainWindow::onDButtonClicked()
{
    setTargetString("D", 146.83f);
}

void MainWindow::onGButtonClicked()
{
    setTargetString("G", 196.00f);
}

void MainWindow::onBButtonClicked()
{
    setTargetString("B", 246.94f);
}

void MainWindow::onE4ButtonClicked()
{
    setTargetString("E4", 329.63f);
}


void MainWindow::updateTunerDisplay(float pitchHz)
{
    if (!recordingActive) return;

    float smoothedHz = applyAdaptiveSmoothing(pitchHz);

    if (smoothedHz > 0.0f) {
        // –û—Ç–æ–±—Ä–∞–∂–∞–µ–º —á–∞—Å—Ç–æ—Ç—É
        ui->frequencyLabel->setText(QString("%1 –ì—Ü").arg(smoothedHz, 0, 'f', 1));

        if (manualStringSelection && currentTargetFrequency > 0) {
            // –†–µ–∂–∏–º —Ä—É—á–Ω–æ–π –Ω–∞—Å—Ç—Ä–æ–π–∫–∏ –Ω–∞ –∫–æ–Ω–∫—Ä–µ—Ç–Ω—É—é —Å—Ç—Ä—É–Ω—É
            float cents = 1200.0f * std::log2(smoothedHz / currentTargetFrequency);

            // –û–±–Ω–æ–≤–ª—è–µ–º –ø—Ä–æ–≥—Ä–µ—Å—Å-–±–∞—Ä
            int barValue = qBound(-50, qRound(cents), 50);
            ui->tuningBar->setValue(barValue);

            // –¶–≤–µ—Ç –∏–Ω–¥–∏–∫–∞—Ü–∏–∏ –≤ –∑–∞–≤–∏—Å–∏–º–æ—Å—Ç–∏ –æ—Ç —Ç–æ—á–Ω–æ—Å—Ç–∏
            QString barColor;
            if (qAbs(cents) < 5) barColor = "#00ff88";
            else if (qAbs(cents) < 10) barColor = "#ffaa00";
            else if (qAbs(cents) < 20) barColor = "#ff5500";
            else barColor = "#ff0000";

            ui->tuningBar->setStyleSheet(
                QString("QProgressBar::chunk { background-color: %1; border-radius: 8px; }").arg(barColor)
                );

            // –û—Ç–æ–±—Ä–∞–∂–∞–µ–º –æ—Ç–∫–ª–æ–Ω–µ–Ω–∏–µ
            ui->centsLabel->setText(QString("Cents: %1").arg(qRound(cents)));
            ui->centsLabel->setStyleSheet(QString("color: %1; background: transparent;").arg(barColor));

            // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º —Ü–µ–ª–µ–≤—É—é –Ω–æ—Ç—É
            ui->noteLabel->setText(QString("%1 ‚ûî %2 –ì—Ü").arg(currentTargetString).arg(currentTargetFrequency, 0, 'f', 1));

        } else {
            // –ê–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∏–π —Ä–µ–∂–∏–º
            QString noteName;
            float targetFreq;
            float cents = NoteConverter::frequencyToCents(smoothedHz, noteName, targetFreq);

            ui->noteLabel->setText(noteName);
            ui->centsLabel->setText(QString("–¶–µ–Ω—Ç—ã: %1").arg(qRound(cents)));

            // –û–±–Ω–æ–≤–ª—è–µ–º –ø—Ä–æ–≥—Ä–µ—Å—Å-–±–∞—Ä
            int barValue = qBound(-50, qRound(cents), 50);
            ui->tuningBar->setValue(barValue);

            // –¶–≤–µ—Ç –∏–Ω–¥–∏–∫–∞—Ü–∏–∏
            QString barColor;
            if (qAbs(cents) < 5) barColor = "#00ff88";
            else if (qAbs(cents) < 10) barColor = "#ffaa00";
            else if (qAbs(cents) < 20) barColor = "#ff5500";
            else barColor = "#ff0000";

            ui->tuningBar->setStyleSheet(
                QString("QProgressBar::chunk { background-color: %1; border-radius: 8px; }").arg(barColor)
                );

            ui->centsLabel->setStyleSheet(QString("color: %1; background: transparent;").arg(barColor));

            // –ê–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–æ–µ –æ–ø—Ä–µ–¥–µ–ª–µ–Ω–∏–µ —Å—Ç—Ä—É–Ω—ã
            highlightCorrectString(smoothedHz);
        }

    } else {
        // –ù–µ—Ç —Å–∏–≥–Ω–∞–ª–∞
        ui->frequencyLabel->setText("--- –ì—Ü");
        ui->noteLabel->setText("---");
        ui->centsLabel->setText("–¶–µ–Ω—Ç—ã: ---");
        ui->tuningBar->setValue(0);
        ui->tuningBar->setStyleSheet("QProgressBar::chunk { background-color: #555555; }");
        ui->centsLabel->setStyleSheet("color: #a0aec0; background: transparent;");

        if (!manualStringSelection) {
            resetStringHighlights();
        }
    }
}

void MainWindow::on_startStopButton_clicked()
{
    if (!recordingActive) {
        audioRecorder->startRecording();
        ui->startStopButton->setText("‚èπ –°—Ç–æ–ø");
        recordingActive = true;
        ui->statusbar->showMessage("–ü—Ä–æ—Å–ª—É—à–∏–≤–∞–Ω–∏–µ...", 2000);
    } else {
        audioRecorder->stopRecording();
        ui->startStopButton->setText("üé§ –°—Ç–∞—Ä—Ç");
        recordingActive = false;
        resetDisplay();
        ui->statusbar->showMessage("–ü—Ä–æ—Å–ª—É—à–∏–≤–∞–Ω–∏–µ –æ—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–æ", 2000);
    }
}

void MainWindow::handleAudioError(const QString& message)
{
    QMessageBox::critical(this, "–û—à–∏–±–∫–∞ –ê—É–¥–∏–æ", message);
    if (recordingActive) {
        audioRecorder->stopRecording();
        ui->startStopButton->setText("üé§ –°—Ç–∞—Ä—Ç");
        recordingActive = false;
        resetDisplay();
    }
}

float MainWindow::applyAdaptiveSmoothing(float newFrequency)
{
    if (newFrequency <= 0.0f) {
        frequencyHistory.clear();
        return 0.0f;
    }

    qint64 currentTime = QDateTime::currentMSecsSinceEpoch();
    FrequencySample sample{newFrequency, currentTime};
    frequencyHistory.append(sample);

    // –£–¥–∞–ª—è–µ–º —Å—Ç–∞—Ä—ã–µ samples (—Å—Ç–∞—Ä—à–µ 500 –º—Å)
    while (!frequencyHistory.isEmpty() &&
           currentTime - frequencyHistory.first().timestamp > 500) {
        frequencyHistory.removeFirst();
    }

    // –ï—Å–ª–∏ —Å–ª–∏—à–∫–æ–º –º–Ω–æ–≥–æ samples, –æ–≥—Ä–∞–Ω–∏—á–∏–≤–∞–µ–º
    if (frequencyHistory.size() > MAX_HISTORY_SIZE) {
        frequencyHistory.removeFirst();
    }

    // –ï—Å–ª–∏ –º–∞–ª–æ –¥–∞–Ω–Ω—ã—Ö, –Ω–µ —Å–≥–ª–∞–∂–∏–≤–∞–µ–º
    if (frequencyHistory.size() < 3) {
        return newFrequency;
    }

    // –í–∑–≤–µ—à–µ–Ω–Ω–æ–µ —Å—Ä–µ–¥–Ω–µ–µ: –±–æ–ª–µ–µ —Å–≤–µ–∂–∏–µ –∑–Ω–∞—á–µ–Ω–∏—è –∏–º–µ—é—Ç –±–æ–ª—å—à–∏–π –≤–µ—Å
    float weightedSum = 0.0f;
    float totalWeight = 0.0f;

    for (int i = 0; i < frequencyHistory.size(); ++i) {
        float age = (currentTime - frequencyHistory[i].timestamp) / 1000.0f; // –í —Å–µ–∫—É–Ω–¥–∞—Ö
        float weight = exp(-age * 2.0f); // –≠–∫—Å–ø–æ–Ω–µ–Ω—Ü–∏–∞–ª—å–Ω–æ–µ –∑–∞—Ç—É—Ö–∞–Ω–∏–µ –≤–µ—Å–∞
        weightedSum += frequencyHistory[i].value * weight;
        totalWeight += weight;
    }

    return weightedSum / totalWeight;
}

void MainWindow::showHelpDialog(){
        QDialog *helpDialog = new QDialog(this);
    helpDialog->setWindowTitle("Tuner Help & Accuracy");
    helpDialog->resize(450, 500);

    helpDialog->setModal(false);
    helpDialog->setAttribute(Qt::WA_DeleteOnClose);

    QVBoxLayout *mainLayout = new QVBoxLayout(helpDialog);
    mainLayout->setContentsMargins(0, 0, 0, 0);

    QScrollArea *scrollArea = new QScrollArea(helpDialog);
    scrollArea->setWidgetResizable(true);
    scrollArea->setFrameShape(QFrame::NoFrame);

    QWidget *contentWidget = new QWidget();
    QVBoxLayout *layout = new QVBoxLayout(contentWidget);
    layout->setSpacing(15);
    layout->setContentsMargins(20, 20, 20, 20);

    // --- –°–µ–∫—Ü–∏—è "–¢–æ—á–Ω–æ—Å—Ç—å" ---
    QLabel *accuracyTitle = new QLabel("–†—É–∫–æ–≤–æ–¥—Å—Ç–≤–æ –ø–æ —Ç–æ—á–Ω–æ—Å—Ç–∏", contentWidget);
    accuracyTitle->setObjectName("section");
    accuracyTitle->setStyleSheet("font-weight: bold; font-size: 16px;");
    layout->addWidget(accuracyTitle);

    QLabel *accuracyText = new QLabel(
        "<table cellspacing='5'>"
        "<tr><td><div style='background-color: #00ff00; width: 20px; height: 20px; border-radius: 3px;'></div></td>"
        "<td><b>–ò–¥–µ–∞–ª—å–Ω–æ</b>: –í –ø—Ä–µ–¥–µ–ª–∞—Ö ¬±5 —Ü–µ–Ω—Ç–æ–≤</td></tr>"
        "<tr><td><div style='background-color: #ffaa00; width: 20px; height: 20px; border-radius: 3px;'></div></td>"
        "<td><b>–•–æ—Ä–æ—à–æ</b>: –í –ø—Ä–µ–¥–µ–ª–∞—Ö ¬±10 —Ü–µ–Ω—Ç–æ–≤</td></tr>"
        "<tr><td><div style='background-color: #ff5500; width: 20px; height: 20px; border-radius: 3px;'></div></td>"
        "<td><b>–°—Ä–µ–¥–Ω–µ</b>: –í –ø—Ä–µ–¥–µ–ª–∞—Ö ¬±20 —Ü–µ–Ω—Ç–æ–≤</td></tr>"
        "<tr><td><div style='background-color: #ff0000; width: 20px; height: 20px; border-radius: 3px;'></div></td>"
        "<td><b>–ü–ª–æ—Ö–æ</b>: –ë–æ–ª–µ–µ ¬±20 —Ü–µ–Ω—Ç–æ–≤</td></tr>"
        "</table><br>"
        "<b>–ü—Ä–∏–º–µ—á–∞–Ω–∏–µ</b>: 100 —Ü–µ–Ω—Ç–æ–≤ = 1 –ø–æ–ª—É—Ç–æ–Ω",
        contentWidget
        );
    accuracyText->setWordWrap(true);
    layout->addWidget(accuracyText);

    // --- –°–µ–∫—Ü–∏—è "–°–æ–≤–µ—Ç—ã" ---
    QLabel *tipsTitle = new QLabel("–°–æ–≤–µ—Ç—ã –ø–æ —ç–∫—Å–ø–ª—É–∞—Ç–∞—Ü–∏–∏", contentWidget);
    tipsTitle->setStyleSheet("font-weight: bold; font-size: 16px;");
    layout->addWidget(tipsTitle);

    QLabel *tipsText = new QLabel(
        "‚Ä¢  –†–µ–∫–æ–º–µ–Ω–¥—É–µ—Ç—Å—è –Ω–∞—Å—Ç—Ä–∞–∏–≤–∞—Ç—å –∏–Ω—Å—Ç—Ä—É–º–µ–Ω—Ç –≤ —Ç–∏—Ö–æ–º –ø–æ–º–µ—â–µ–Ω–∏–∏<br>"
        "‚Ä¢  –î–µ—Ä–∂–∞—Ç—å –Ω–æ—Ç—É 2-3 —Å–µ–∫—É–Ω–¥—ã –¥–ª—è –±–æ–ª–µ–µ —Ç–æ—á–Ω—ã—Ö —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤<br>"
        "‚Ä¢  –ù–µ –∑–∞–¥–µ–π—Å—Ç–≤–æ–≤–∞—Ç—å –±–æ–ª–µ–µ –æ–¥–Ω–æ–π –Ω–æ—Ç—ã<br>"
        "‚Ä¢  –ò–Ω—Ç–µ—Ä—Ñ–µ–π—Å —Ç—é–Ω–µ—Ä–∞ —É–¥–æ–±–µ–Ω –¥–ª—è –Ω–∞—Å—Ç—Ä–æ–π–∫–∏ –≥–∏—Ç–∞—Ä—ã, –Ω–æ –ø—Ä–æ–≥—Ä–∞–º–º—É –º–æ–∂–Ω–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å –¥–ª—è –Ω–∞—Å—Ç—Ä–æ–π–∫–∏ –¥—Ä—É–≥–∏—Ö –º—É–∑—ã–∫–∞–ª—å–Ω—ã—Ö –∏–Ω—Å—Ç—Ä—É–º–µ–Ω—Ç–æ–≤ –µ—Å–ª–∏ —Å—Ä–∞–≤–Ω–∏–≤–∞—Ç—å –æ–∂–∏–¥–∞–µ–º—É—é –Ω–æ—Ç—É —Å –≤—ã–≤–æ–¥–∏–º–æ–π<br>"
        ""
        "‚Ä¢  –°—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–π —Å—Ç—Ä–æ–π —à–µ—Å—Ç–∏—Å—Ç—Ä—É–Ω–Ω–æ–π –≥–∏—Ç–∞—Ä—ã:<br>"
        "       E2: 82.41 –ì—Ü | A: 110.00 –ì—Ü | D: 146.83 –ì—Ü<br>"
        "       G: 196.00 –ì—Ü | B: 246.94 –ì—Ü | E4: 329.63 –ì—Ü<br>",
        contentWidget
        );
    tipsText->setWordWrap(true);
    layout->addWidget(tipsText);

    // --- –°–µ–∫—Ü–∏—è "–ù–æ—Ç—ã" ---
    QLabel *notesTitle = new QLabel("–ë—É–∫–≤–µ–Ω–Ω–æ–µ –æ–±–æ–∑–Ω–∞—á–µ–Ω–∏–µ –Ω–æ—Ç –∏ —Ç–æ–Ω–∞–ª—å–Ω–æ—Å—Ç–µ–π", contentWidget);
    notesTitle->setStyleSheet("font-weight: bold; font-size: 16px;");
    layout->addWidget(notesTitle);

    QLabel *notesText = new QLabel(
        "‚Ä¢  –ë—É–∫–≤–µ–Ω–Ω—ã–º –æ–±–æ–∑–Ω–∞—á–µ–Ω–∏—è–º, –≤—ã–≤–æ–¥–∏–º—ã–º –ø—Ä–æ–≥—Ä–∞–º–º–æ–π, —Å–æ–æ—Ç–≤–µ—Ç—Å—Ç–≤—É—é—Ç —Å–ª–µ–¥—É—é—â–∏–µ —Å–ª–æ–≥–æ–≤—ã–µ:<br>"
        "   C - –î–æ<br>"
        "   D - –†–µ<br>"
        "   E - –ú–∏<br>"
        "   F - –§–∞<br>"
        "   G - –°–æ–ª—å<br>"
        "   A - –õ—è<br>"
        "   B - –°–∏<br>"
        ""
        "‚Ä¢  –¶–∏—Ñ—Ä–∞ –ø–æ—Å–ª–µ –Ω–æ—Ç—ã –æ–±–æ–∑–Ω–∞—á–∞–µ—Ç –æ–∫—Ç–∞–≤—É (–ù–∞–ø—Ä–∏–º–µ—Ä: E2 - –ú–∏ –≤—Ç–æ—Ä–æ–π –æ–∫—Ç–∞–≤—ã)"
        "‚Ä¢  –î–∏–µ–∑ (‚ôØ) ‚Äî —ç—Ç–æ –∑–Ω–∞–∫, –∫–æ—Ç–æ—Ä—ã–π –ø–æ–≤—ã—à–∞–µ—Ç –Ω–∞—Ç—É—Ä–∞–ª—å–Ω—É—é –Ω–æ—Ç—É –Ω–∞ –ø–æ–ª—Ç–æ–Ω–∞ (100 —Ü–µ–Ω—Ç–æ–≤) –≤–≤–µ—Ä—Ö",
        contentWidget
        );
    notesText->setWordWrap(true);
    layout->addWidget(notesText);

    // –ö–Ω–æ–ø–∫–∞ –∑–∞–∫—Ä—ã—Ç–∏—è
    QPushButton *closeButton = new QPushButton("–Ø—Å–Ω–æ", contentWidget);
    closeButton->setMinimumHeight(35);
    connect(closeButton, &QPushButton::clicked, helpDialog, &QDialog::close);
    layout->addWidget(closeButton, 0, Qt::AlignCenter);

    layout->addStretch();

    scrollArea->setWidget(contentWidget);
    mainLayout->addWidget(scrollArea);

    helpDialog->show();
}
